### ISSUE TITLE
Genome Extractor

### ISSUE THREAD

**Issue by BillMills** [goyalsid/phageParser#3]
As described in #2, we want to know more about a phage than just what genes matched our query; we want to know what sequences surround that match too.  In #2 we described some functions for fetching complete genome datasets from genbank; now we need to extract the relevant information from those results.

We need a python function `findNeighbours(startPosition, endPosition, n, m)` that examines the file downloaded via `prefetch` (see #2), and returns a string corresponding to the genome starting `n` bases before `startPosition` and carrying on to `m` bases after `endPosition`.  If either `n` or `m` are not provided, default them to 100.

**Comment by mbonsma** [comment]
This is biologically relevant to the question of [Protospacer Adjacent Motifs](http://en.wikipedia.org/wiki/Protospacer_adjacent_motif) (PAMs), which are short sequences in the vicinity of viral genome regions that become spacers. Some systems seem to have them, some don't. It will be interesting to have a way to search all hits to viral genomes for PAMs, which this function would provide.

**Comment by mbonsma** [comment]
Here is an [example phage genome file](https://github.com/goyalsid/phageParser/blob/master/data/NC_020879_phage.gb) that would be returned from `acc2gb.py`. The function described in this issue would use this example file as input.

**Comment by neaGaze** [comment]
Hi @mbonsma ! I'm trying to understand the issue here. Can you help me with that? Since I have a very limited knowledge in biology, I think it might be much easier if you could present an example of the function`findNeighbours(startPosition, endPosition, n, m)`. For Eg: in this [input file](https://github.com/goyalsid/phageParser/blob/master/data/NC_020879_phage.gb) what portion of the data do we really care about? Can you please give a test case giving some arbitrary parameter at startPostion, endPostion, n and m values with its respective answer? This way I can better understand what exactly needs to be done.

With Regards

**Comment by mbonsma** [comment]
@neaGaze, the part of the file we care about in this case is the DNA sequence which starts [here](https://github.com/goyalsid/phageParser/blob/master/data/NC_020879_phage.gb#L4049). 

Here's an example:
```
startPosition = 181
endPosition = 240
n = 60
m = 60
```

This would return these lines from the file (minus the base number), starting at [base 121](https://github.com/goyalsid/phageParser/blob/master/data/NC_020879_phage.gb#L4051) and ending with base 299 in [this line](https://github.com/goyalsid/phageParser/blob/master/data/NC_020879_phage.gb#L4053).


```
      121 catttcagaa tactttttat tctgttcatc aacatattct ttgatttcag caaaatcggc
      181 ataatcggga tttccgattt caacccattg ggatgcatca ttcaggaaat catcattgct
      241 ttcgtcaaac ttttcggtta tatcaattcc aagaatgccc atttgcttaa gacgaattgc
```

You can use python-style indexing if you prefer (aka `endPosition = 241` instead of 240), as long you indicate which you chose!

**Comment by neaGaze** [comment]
Hi @mbonsma  I have submitted the patch. Could you please check them if they are correct? Thanks!

**Comment by mbonsma** [comment]
Thanks @neaGaze, I will take a look this weekend! :+1:

**Comment by mbonsma** [comment]
@neaGaze, it looks good! Just a couple things:

- right now the four input numbers (`startPosition, endPosition, n, m`) are hardcoded on line 18 of `getGenomes.py`, and since these will change each time it's used, could you make them input arguments instead?
- can you add a short docstring to the getGenomes.py function describing usage? 
- can you open a pull request with your changes to the main repo? Then I can merge it.

Thanks!

**Comment by neaGaze** [comment]
HI @mbonsma Thanks for reviewing. `getGenomes.py` is just an example of how the feature can be used. The real thing is inside the folder: `parsers/genome_extractor.py`. I have written some docstring in that file including the usage. I believe that I have shown the usage instruction in that file too, however if you believe that there is a need for some explanation beside that please let me know. Lastly, I think I have already sent the pull request in the main repo too.

**Comment by mbonsma** [comment]
Hi @neaGaze, sorry for the very slow reply! [Here's](https://github.com/goyalsid/phageParser/pulls) where the pull request should end up - I don't see it there, perhaps you can link to your PR?

**Comment by neaGaze** [comment]
Yaa I'm sorry! I thought the request was submitted but I was wrong. Now
it's done I guess.

Regards,
Nigesh


::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::



[image: --]

Nigesh Shakya
[image: https://]about.me/neaGaze
 <https://about.me/neaGaze?promo=email_sig&utm_source=email_sig&utm_medium=email_sig&utm_campaign=external_links>


On Tue, Jun 28, 2016 at 10:18 AM, Madeleine Bonsma <notifications@github.com

**Comment by mbonsma** [comment]
Fixed in PR #94.
